espnet2/train/deepspeed_trainer.py

```diff
@@ -0,0 +1,273 @@
+            model_parameters=model.parameters(),
+            config=deepspeed_config,
+        )
+
+        # (3) setup reporter, output_dir, dataloader etc.
+        output_dir = Path(trainer_options.output_dir)
+        reporter = Reporter()
+
+        # (4) resume
+        if trainer_options.resume:
+            cls.resume(
+                model=model,
+                reporter=reporter,
+                output_dir=output_dir,
+            )
+
+        # (5) loop on epochs
+        start_epoch = reporter.get_epoch() + 1
+        if start_epoch == trainer_options.max_epoch + 1:
+            logging.warning(
+                f"The training has already reached at max_epoch: {start_epoch}"
+            )
+
+        for iepoch in range(start_epoch, trainer_options.max_epoch + 1):
+            set_all_random_seed(trainer_options.seed + iepoch)
+            reporter.set_epoch(iepoch)
+
+            # (5.1) train one epoch
+            with reporter.observe("train") as sub_reporter:
+                cls.train_one_epoch(
+                    model=model,
+                    iterator=train_iter_factory.build_iter(iepoch),
+                    reporter=sub_reporter,
+                    options=trainer_options,
+                )
+
+            # (5.2) valid one epoch
+            with reporter.observe("valid") as sub_reporter:
+                cls.valid_one_epoch(
+                    model=model,
+                    iterator=valid_iter_factory.build_iter(iepoch),
+                    reporter=sub_reporter,
+                    options=trainer_options,
+                )
+
+            # (5.3) save checkpoint
+            checkpoint_path = output_dir / f"checkpoint_{iepoch}"
+            model.save_checkpoint(
+                checkpoint_path,
+                tag=f"{iepoch}",
+                client_state={"reporter": reporter.state_dict()},
+            )
+
+            # (5.4) reporter
+            if dist.get_rank() == 0:
+                logging.info(reporter.log_message())
+                reporter.matplotlib_plot(output_dir / "images")
+
+    @classmethod
+    @typechecked
+    def train_one_epoch(
+        cls,
+        model,
+        iterator: Iterable[Tuple[List[str], Dict[str, torch.Tensor]]],
+        reporter: SubReporter,
+        options: DeepSpeedTrainerOptions,
+    ) -> None:
+        model.train()
+        iterator_stop = torch.tensor(0).cuda()
+
+        log_interval = options.log_interval
+        if log_interval is None:
+            try:
+                log_interval = max(len(iterator) // 20, 10)
+            except TypeError:
+                log_interval = 100
+
+        for iiter, (utt_id, batch) in enumerate(
+            reporter.measure_iter_time(iterator, "iter_time"), 1
+        ):
+            assert isinstance(batch, dict), type(batch)
+
+            with reporter.measure_time("step_time"):
+                # (0) ensure all ranks have not finished.
+                dist.all_reduce(iterator_stop, ReduceOp.SUM)
```

Comment:
This all-reduce is not necessary (same with the one in `valid_one_epoch`). We can remove lines 185-187

Reply:
I think that forced synchronization is necessary, as it functions like a join operation, allowing all ranks to start forwarding simultaneously and avoiding timeouts. Torch_DDP does not require this synchronization operation, partly because DDP has its own join function, and partly because, for distributed training like DeepSpeed, the communication between machines is more complex and timeouts are more likely to occur. Therefore, sacrificing some waiting time (in fact, I believe the proportion of waiting time will not be very high) in exchange for training stability is necessary.

Reply:
Thanks for the both comments :)
Sice the ZeRO will always have extensive communication, maybe this force sync is ok to use. I'll post more observation if I find these lines are the performance bottleneck.